README.md
<p align="center">
  <a href="https://www.phpdish.com/">
    <img alt="PHPDish" width="400" src="https://raw.githubusercontent.com/slince/phpdish/master/assets/img/logo2.png"/>
  </a>
</p>

PHPDish 是一个基于Symfony框架开发的内容社区系统；得益于大量的前端以及后端的第三方类库的使用使得PHPDish有着高质量的代码，敏捷实现；
你可以使用composer或者直接下载本仓库进行程序的安装，注意切换到tag。
QQ群号：138307655，欢迎进群讨论。

[PHPDish 开发手册](https://www.phpdish.com/books/docs)

## Requirements:

- PHP 7.2.0+
- MYSQL 5.6.0+

## Features

- 社区话题/回复
- 专栏文章/评论
- Notification通知
- Markdown编辑器
- Github风格Emoji表情支持
- 在回复中@他人
- 个人主页
- 电子书编辑/阅读
- 专栏/电子书付费订阅
- 钱包管理

关于部分功能的使用以及实现请关注”[PHPDish开发手记](https://www.phpdish.com/categories/phpdish-development-notes)“专栏，我会在该专栏内记录一些开发经历与使用方法；

PHPDish功能仍在继续实现中，陆续会有新的功能发布出来，欢迎关注，欢迎贡献代码。如果你有好的建议或者问题需要反馈,
可以在[Issue](https://github.com/slince/phpdish/issues) 发布新贴；或者到PHPDish社区 [http://www.phpdish.com](http://www.phpdish.com) 
发布新的话题。如果没有及时回复你可以在PHPDish私信我[@slince](http://www.phpdish.com/users/slince)；

## 文档

安装文档查看这里 [/app/Resources/doc/install.md](./app/Resources/doc/install.md)

## 开源库

项目中使用的第三方库比较多不能一一列举，这里提一些比较核心功能使用到的库

- 后端

| 名称 | 说明 |
| --- | --- |
| [friendsofsymfony/user-bundle](https://github.com/FriendsOfSymfony/FOSUserBundle) | 用户管理基础组件 |
| [friendsofsymfony/message-bundle](https://github.com/FriendsOfSymfony/FOSMessageBundle) | 站内信实现的基础组件 |
| [knplabs/knp-markdown-bundle](https://github.com/KnpLabs/KnpMarkdownBundle) | Markdown 解析组件 |
| [knplabs/knp-gaufrette-bundle](https://github.com/KnpLabs/KnpGaufretteBundle) | 文件系统管理组件 |
| [knplabs/knp-menu-bundle](https://github.com/KnpLabs/KnpMenuBundle) | 菜单栏生成组件 |
| [hwi/oauth-bundle](https://github.com/hwi/HWIOAuthBundle) | OAuth登录 |
| [emojione/emojione](https://github.com/emojione/emojione) | Emoji 表情解析的后端组件 |
| [lincanbin/material-design-avatars](https://github.com/lincanbin/Material-Design-Avatars) | 字符头像生成库 |

- 前端

| 名称 | 说明 |
| --- | --- |
| [jquery-pjax](https://github.com/defunkt/jquery-pjax) | PJax |
| [marked](https://github.com/chjj/marked) | Markdown 解析组件，前端用 |
| [plupload](https://github.com/moxiecode/plupload) | 文件上传 |
| [twemoji](https://github.com/twitter/twemoji) | Emoji 解析 |
| [textcomplete](https://github.com/yuku-t/textcomplete) | 文本域内自动完成，@用户和emoji提示由此组件实现 |
| [store](https://github.com/marcuswestin/store.js) | 数据存储组件，可将数据写入到多个媒介 |
| [highlight.js](https://github.com/isagalaev/highlight.js) | 正文内代码高亮 |
| [inline-attachment](https://github.com/Rovak/InlineAttachment) | 粘贴板复制粘贴上传以及文件拖拽上传 |
| [codemirror](https://github.com/codemirror/CodeMirror) | 前端编辑器 |
| [simplemde](https://github.com/sparksuite/simplemde-markdown-editor) | Markdown编辑器 |

## 赞助 Donate

开源项目的发展离不开大家的支持，如果项目对你有所帮助你可以赞助我帮助我把 PHPDish 做的更好。

- 微信/支付宝

<img src="https://raw.githubusercontent.com/slince/phpdish/master/app/Resources/assets/alipay.png" height="350"/><img src="https://raw.githubusercontent.com/slince/phpdish/master/app/Resources/assets/wechat-pay.png" height="350"/>

- PayPal

[https://www.paypal.me/slince](https://www.paypal.me/slince)

捐赠时请留下您的 GitHub 或者个人主页等相关的个人信息 :heart:。

### Donate 列表

| 赞助者 | 赞助金额 (￥)|
| --- | --- |
| [Intern](https://www.xde.io) | 50 |
| 阳阳 | 10 |
| [アクア様が見てる@沟槽](https://www.phpdish.com/users/gouchaoer) | 200 + 200 |
| 未知用户 | 3 |

## 联系我

- 邮箱：taosikai@yeah.net
- Github： https://github.com/slince
- Gitee： https://gitee.com/slince
- 微信：

<img src="https://raw.githubusercontent.com/slince/phpdish/master/app/Resources/assets/wechat.jpg" width="200"/>


## CHANGELOG

更新日志 [CHANGELOG](./CHANGELOG.md)

## License

PHPDish 采用 [MIT](https://opensource.org/licenses/MIT) 开源许可证，你可以在商业项目中免费使用 PHPDish 或者基于 PHPDish 二次
开发而不必支付费用。


CHANGELOG.md
# Changelog

## 0.3.3 - 2018-05-17

Fixed:

- 重置版本号
- 修复前端样式
- 允许中文名注册

Added:

- 开放后台用户编辑操作
- 增加简繁英三种语言支持

## 0.3.2 - 2018-04-02

Fixed:

- 修复 algolia 没有配置无法安装的问题

## 0.3.1 - 2018-03-20

Fixed:

- 修复电子书阅读目录显示不正确的问题
- 修复电子书页面错误
- 修改文案
- 调整有赞支付sdk为beta版
- 修复用户的电子书与话题数量字段不准确的问题
- 修复艾特楼主楼主会收到提醒的问题
- 修复部分接口暴露过多字段的问题

Added:

- 增加模板函数，允许在模板里直接获取相关数据
- 增加多主题模板支持
- 增加插件支持
- 数据模型层优化，将实体元格式从 annotation 调整为 xml
- 增加支持电子书目录顺序调整
- 新增主题列表打印命令与主题资源安装命令
- 优化部分 bundle 的别名
- 从核心代码抽离每日一首歌的功能到插件，如果需要的话安装该插件即可
- 增加找回密码流程
- 增加话题/回复/文章/评论点赞功能
- 优化文章页数据库查询
- 提取注册提醒的内容和发信人到参数配置
- 优化搜索业务逻辑，为实现在插件中提供搜索服务做基础
- 迁移升级 algolia 搜索依赖到最新版本
- 优化代码结构，将大多数不会变的代码提取到底层


## 0.2.1 - 2018-03-05

Fixed:

- 修复前端草稿对话题/文章/电子书编辑影响
- 修复新版本symfony导致的密码修改bug


Added:

- PHP 版本最低为7.2.0
- Symfony 版本最低为3.4.0
- 更新已经废除的第三方包
- 允许修改用户名
- 帖子置顶支持
- 用户主页关注按钮判断增强


## 0.1.2 - 2018-03-01

- 修复预览没有高亮的问题
- 修复 Markdown 样式
- 更换图标
- 过滤用户提交内容中的标签
- 免费订阅的专栏/电子书不再显示购买按钮


## 0.1.1 - 2018-02-22
- 增加钱包功能
- 增加电子书功能
- 增加专栏/电子书付费订阅购买功能
- 全新的后台管理

## 0.0.2 - 2018-01-11
- 修复@中文用户名失效的问题
- Threads标签化，发帖时自动补全
- 增加节点编辑/关注功能
- 增加页面右侧数据缓存
- 修复对话列表页不分页的问题
- 修复话题回复内容过长导致提醒出错的bug
- 增加浏览器Html5通知


assets/js/book.js
import '../modules/common.js';
import hljs from 'highlight.js';
import SocialShare from 'social-share-button.js';
import NProgress from 'nprogress';
import 'jquery-pjax';
import AjaxTab from "../modules/ajaxtab";
import Util from "../modules/util";
import {FollowUserIntialization} from "../modules/actions";
import lockButton from '../modules/button-lock.js';
import InlineAttachment from "../modules/inline-attachment";
import SimpleMDE from "simplemde";
import QRCodePayment from "../modules/qrcode-payment";
import md5 from 'blueimp-md5';

//电子书详情页面
const $bookDetails = $('#book-details');
$bookDetails.length > 0 && (function($){
    new AjaxTab($('[data-pjax-container]'), {
        container: '#book-details',
        loader: '#loader',
        before: (container) => {
            Util.htmlPlaceholder(container);
        },
        success: (container) => {
            new FollowUserIntialization(container);
            initBookSummary();
        }
    });
    function initBookSummary(){
        const $addChapter = $bookDetails.find('[data-role="add-chapter"]');
        const btnLock = lockButton($addChapter);
        $addChapter.on('click', function(){
            if (btnLock.isDisabled()) {
                return false;
            }
            btnLock.lock();
            Util.dialog.inputs(Translator.trans('book.chapter_name'), [{name: 'title', required: true}], {
                messages: {
                    title: {
                        "required": Translator.trans('book.required_chapter_name')
                    }
                }
            }, {
                'okValue': Translator.trans('ui.create'),
                'cancelValue': Translator.trans('ui.cancel'),
                width: 350
            }).then((data)=>{
                console.log(data);
                Util.request('book.add_summary', {slug: window.book.slug}, data).done(()=>{
                    location.reload();
                }).fail(()=>{
                    Util.dialog.message(Translator.trans('book.create_chapter_error')).flash();
                });
                btnLock.release();
            }, ()=>{
                btnLock.release();
            });
        });
        //处理action
        const $bookChapters = $bookDetails.find('[data-role="chapter"]');
        $bookChapters.each(function(){
            const $this = $(this);
            const $edit = $this.find('[data-role="edit"]');
            const $addSub = $this.find('[data-role="add-sub"]');
            const $delete = $this.find('[data-role="delete"]');
            const $move = $this.find('[data-role="move"]');

            $edit.on('click', function(){
                Util.dialog.inputs(Translator.trans('book.chapter_name'), [{
                    name: 'title',
                    default: $this.data('title'),
                    required: true
                }], {
                    messages: {
                        title: {
                            "required": Translator.trans('book.required_chapter_name')
                        }
                    }
                }, {
                    'okValue': Translator.trans('ui.create'),
                    'cancelValue': Translator.trans('ui.cancel'),
                    width: 350
                }).then((data)=>{
                    Util.request('book.edit_summary', {slug: window.book.slug, id: $this.data('id')}, data).done(()=>{
                        location.reload();
                    }).fail(()=>{
                        Util.dialog.message(Translator.trans('book.edit_chapter_error')).flash();
                    });
                }, ()=>{
                });
            });
            const deleteLock = lockButton($delete);
            $delete.on('click', ()=>{
                "use strict";
                if (deleteLock.isDisabled()) {
                    return false;
                }
                deleteLock.lock();
                Util.dialog.confirm(Translator.trans('book.confirm_remove_chapter')).then(()=>{
                    let $chapter =  $delete.closest('[data-role="sub-chapter"]');
                    if ($chapter.length === 0) {
                        $chapter = $delete.closest('[data-role="chapter"]');
                    }
                    const chapterId =  $chapter.data('id');
                    Util.request('post.delete', chapterId).done(()=>{
                        Util.dialog.message(Translator.trans('book.remove_success')).flash(()=>{
                            location.reload();
                        });
                    }).fail(()=>{
                        Util.dialog.message(Translator.trans('book.remove_error')).flash();
                    }).always(()=>{
                        deleteLock.release();
                    });
                }, ()=>{
                    deleteLock.release();
                });
            });
            const moveLock = lockButton($move);
            $move.on('click', function(){
                const $this = $(this);
                "use strict";
                if (moveLock.isDisabled()) {
                    return false;
                }
                moveLock.lock();
                let $chapter = $this.closest('[data-role="sub-chapter"]');
                if ($chapter.length === 0) {
                    $chapter = $this.closest('[data-role="chapter"]');
                }
                const chapterId =  $chapter.data('id');
                Util.request('book.move_chapter', {slug: window.book.slug, id: chapterId}, {
                    'direction': $(this).data('direction') || 'up',
                    'step': 1
                }).done(()=>{
                    location.reload();
                }).fail(()=>{
                    Util.dialog.message(Translator.trans('book.move_error')).flash();
                }).always(()=>{
                    moveLock.release();
                });
            });
        });
    }
    initBookSummary();
})($);


//添加章节
const chapterBody = document.getElementById('chapter_originalBody');
const $chapterBody = $(chapterBody);
$chapterBody.length > 0 && (function($){
    const $postTitle = $('#chapter_title');
    const $addChapterForm = $('#add-chapter-form');
    const $addChapterBtn = $('[data-action="add-chapter"]');

    const simplemde = new SimpleMDE({
        element: chapterBody,
        autofocus: true,
        spellChecker: false,
        status: false,
        indentWithTabs: false,
        tabSize: 4,
        autosave: {
            enabled: true,
            uniqueId: 'chapter_' + md5(location.pathname),
            delay: 1000,
        },
        toolbar: [
            "bold", "italic", "heading", "|", "quote", "code", "table",
            "horizontal-rule", "unordered-list", "ordered-list", "|",
            "link", "image", "|",  "side-by-side", "fullscreen", "preview", "|",
            {
                name: 'guide',
                action: 'https://github.com/riku/Markdown-Syntax-CN/blob/master/syntax.md',
                className: 'fa fa-info-circle',
                title: Translator.trans('editor.markdown_synax'),
            }
        ],
    });
    new InlineAttachment(simplemde.codemirror); //处理附件上传的功能
    $addChapterBtn.on('click', () => {
        if ($postTitle.val().length === 0) {
            Util.dialog.message(Translator.trans('book.validation.name')).flash();
            return false;
        }
        const buttonLock = lockButton($addChapterBtn).lock();
        Util.dialog.confirm(Translator.trans('book.confirm_publish')).then(()=> {
            $addChapterForm.submit();
            return true;
        }, () => {
            buttonLock.release();
            return false;
        });
        return false;
    });

    //添加文章验证
    $addChapterForm.validate({
        rules: {
            'chapter[title]': {
                required: true,
                rangelength: [2,50]
            }
        },
        messages: {
            'chapter[title]': {
                required: Translator.trans('book.chapter.validation.title.required'),
                rangelength: Translator.trans('book.chapter.validation.title.length_between')
            },
        }
    });
})($);



//电子书阅读页面
const $bookView = $('[data-role="book-view"]');
$bookView.length > 0 && (function($){
    const $bookSummary = $bookView.find('[data-role="summary"]');
    const $summaryToggleBtn = $bookView.find('[data-role="toggle-summary"]');
    $summaryToggleBtn.on('click', function(){
        $bookView.toggleClass('with-summary');
    });
    //分享
    new SocialShare($bookView.find('[data-role="social-share"]'), {
        'theme': 'dark-square',
        'facebook': false,
        'twitter': false
    });
    //代码高亮
    $('pre code').each(function (i, block) {
        hljs.highlightBlock(block);
    });

    (function($){
        const $document = $(document);
        const $characters = $bookSummary.find('li.chapter');
        $.pjax.defaults.timeout = 50000;
        $(document).pjax('ul.summary li a', '#pjax-container')
        $document.on('pjax:start', function() {
            NProgress.start();
        });
        $document.on('pjax:end', function(event) {
            if (event.relatedTarget) {
                const $relatedTarget = $(event.relatedTarget);
                $characters.removeClass('active');
                $bookSummary.find('li.sub-chapter-item').removeClass('active');
                const $subCharacters = $relatedTarget.closest('.sub-chapter');
                if ($subCharacters.length > 0) {
                    $relatedTarget.closest('.sub-chapter-item').addClass('active');
                } else {
                    $relatedTarget.closest('.chapter').addClass('active');
                }
            }
            //代码高亮
            $('pre code').each(function (i, block) {
                hljs.highlightBlock(block);
            });
            NProgress.done();
        });
    })($);

    //购买
    const $buy = $('[data-role="buy"]');
    $buy.length > 0 && (function(){
        const buttonLock = lockButton($buy);
        const slug = $buy.data('slug');
        $buy.on('click', function(){
            const wait = Util.dialog.wait.ballPulse();
            Util.request('category.follow', {'slug': slug}).done(function(response){
                if (response.require_payment) {
                    new QRCodePayment(response.qrcode);
                    return;
                } else {
                    location.reload();
                }
            }).fail(function(response){
                Util.dialog.message(response.responseJSON.error).flash();
            }).always(() => {
                wait.close();
                buttonLock.release();
            });
        });
    })($);
})($);